### PR TITLE
AUI-1908 Tabbing to modal from outside does not work as expected when it is initialized.

### PR DESCRIPTION
**Pull request by zxdgoal**
Hey Maira

We got inconsistent behaviour for tabbing into modal: The first time we can tab into the modal's field; but the second time cannot.

Please have a look at [LPS-55692](https://issues.liferay.com/browse/LPS-55692) and [AUI-1908](https://issues.liferay.com/browse/AUI-1908).

This issue is mirror, but we do receive this kinds of issues from our Japanese customer.

Also, could you give your reasons and concerns about this [pr](https://github.com/liferay/yui3/commit/7a5f777a1361fe5d3325ef30f9d5f20f120b4f83) you added before so that we can explain to our customer why the behaviour of preventing foucsing on modal from outside is intended ?

Thanks
John.

### CODE REVIEW

**Comment by zxdgoal** [comment]
@mariatma

Bump

**Comment by zxdgoal** [comment]
@mairatma

Bump

**Comment by mairatma** [comment]
Hey @zxdgoal, I'm sorry for the delay here.
First, let me explain the pr you asked about. The idea of the pr is not to prevent focus on the modal contents, but to prevent focus on elements that are outside the modal when it's open. That's expected behavior for modal dialogs, they are supposed to block any content behind the backdrop, as if they are the only content on the screen, so this code was fixing a bug where tabs were still able to access those elements.
I'm not sure I understand what the current issue is though. I've read both tickets but it's still confusing to me. Could you send me a gif or a picture that shows the bug? The wording seems to suggest that it's something to do with tabbing to the inside of the modal, but that should be allowed, it just shouldn't be allowed to tab to the outside, like I explained. Can you explain better?

**Comment by zxdgoal** [comment]
Hey Maira

Sorry for the confusion. 

Let me rephrase the issues

In the first time of clicking the select button to open a dialogue, we actually initialize one with a template. 

And this time, we can focus on the modal with tab key from outside ( the X on the right top of dialog in our example).  This is not expected. 

But after you close the dialog and re-click on to button to open the dialogue (Just show a hidden dialogue). We call the dialogue.show() method. And this time, you can not use tab key to focus on modal from outside. This is intended behavior. 

The reason why the first time we can focus on modal from outside is that the block method actually is called before a iframe is present in DOM. So it cannot apply the logic.

So  we can change the iframe template with this logic so that it cannot be foused after initialization. 

And in this way we will get consistent behaviour: for the first time and second time or more, we cannot use the tab key to focus from outside.

**Comment by mairatma** [comment]
What I don't understand is why we don't want people to be able to focus on the modal from outside. That code wasn't intended to do that. When I added it the idea was the opposite as I explained, to not allow focusing on elements outside of the modal. Why don't we want it to be possible for people to tab inside the modal after it's opened?

**Comment by zxdgoal** [comment]
The code could not work as expected, when we open the dialog at second time. pressing the tab key continuously and it will stop at X and will not foucs on the modal because it's tabindex equals -1.

**Comment by zxdgoal** [comment]
I wonder if we should consider the X is something outside. If yes, it should not be focused any way.

**Comment by mairatma** [comment]
I think it should be fine to focus on the X, it actually looks important that it should be focusable to me, since it helps accessibility. We really just don't want people focusing to elements that are covered by the backdrop.

If I understood the problem correctly now,  then the bug seems to be that the contents of the modal are not focusable in that case. Is that because the modal contains an iframe? We need to make sure that elements behind the modal's backdrop are never focusable, but elements showing up in front (which are the modal contents) are always focusable.

**Comment by mairatma** [comment]
It would make sense for this to be happening if the modal has an iframe. If that's the case then the code needs some kind of check to only block iframes that are outside the modal, so that would be the fix here.

**Comment by zxdgoal** [comment]
You are right. The modal has an iframe and setting iframe's tabindex to be -1 to stop it being focused. 

I did reverse the indented and unexpected baahvoir. 

For each time, we need to be able to focuse iframe and the X with tab key. 

I need investigate the code and try to finger out a fix in the direction you point out. 

Many thanks

**Comment by mairatma** [comment]
Sure, let me know if you need any help :)

**Comment by zxdgoal** [comment]
Hey @mairatma

I sent a pr [here](https://github.com/eduardolundgren/yui3/pull/37#).

Please help review it.

Thanks
John
